velox/functions/prestosql/aggregates/benchmarks/ReduceAgg.cpp

```diff
@@ -0,0 +1,232 @@
+    }
+
+    filePath_ = TempFilePath::create();
+    writeToFile(filePath_->path, vectors);
+  }
+
+  ~ReduceAggBenchmark() override {
+    HiveConnectorTestBase::TearDown();
+  }
+
+  void TestBody() override {}
+
+  static inline const std::string kSum = "sum(i16)";
+  static inline const std::string kReduceAgg =
+      "reduce_agg(i16, 0, (x, y) -> (x + y), (x, y) -> (x + y))";
+
+  void verify() {
+    // Verify global aggregation.
+    {
+      auto sumResults = runOnce(makeGlobalPlan(kSum));
+      auto reduceAggResults = runOnce(makeGlobalPlan(kReduceAgg));
+
+      assertEqualResults({sumResults}, {reduceAggResults});
+    }
+
+    // Verify group by aggregation.
+    {
+      auto sumResults = runOnce(makeGroupByPlan(kSum));
+      auto reduceAggResults = runOnce(makeGroupByPlan(kReduceAgg));
+
+      assertEqualResults({sumResults}, {reduceAggResults});
+    }
+  }
+
+  void runGlobalSum() {
+    runGlobal(kSum);
+  }
+
+  void runGroupBySum() {
+    runGlobal(kSum);
+  }
+
+  void runGlobalReduceAgg() {
+    runGlobal(kReduceAgg);
+  }
+
+  void runGroupByReduceAgg() {
+    runGroupBy(kReduceAgg);
+  }
+
+ private:
+  RowVectorPtr runOnce(const core::PlanFragment& planFragment) {
+    auto task = makeTask(planFragment);
+
+    std::vector<VectorPtr> results;
+    while (auto result = task->next()) {
+      results.push_back(result);
+    }
+
+    return copyResults(planFragment.planNode->outputType(), results);
+  }
+
+  RowVectorPtr copyResults(
+      const TypePtr& resultType,
+      const std::vector<VectorPtr>& results) {
+    if (results.empty()) {
+      return BaseVector::create<RowVector>(resultType, 0, pool());
+    }
+
+    auto totalCount = 0;
+    for (const auto& result : results) {
+      totalCount += result->size();
+    }
+
+    auto copy = BaseVector::create<RowVector>(resultType, totalCount, pool());
+    auto copyCount = 0;
+    for (const auto& result : results) {
+      copy->copy(result.get(), copyCount, 0, result->size());
+      copyCount += result->size();
+    }
+
+    return copy;
+  }
+
+  void runGlobal(const std::string& aggregate) {
+    folly::BenchmarkSuspender suspender;
+
+    auto plan = makeGlobalPlan(aggregate);
+    auto task = makeTask(plan);
+
+    suspender.dismiss();
+
+    vector_size_t numResultRows = 0;
+    while (auto result = task->next()) {
+      numResultRows += result->size();
+    }
+
+    folly::doNotOptimizeAway(numResultRows);
+  }
+
+  void runGroupBy(const std::string& aggregate) {
+    folly::BenchmarkSuspender suspender;
+
+    auto plan = makeGroupByPlan(aggregate);
+    auto task = makeTask(plan);
+
+    suspender.dismiss();
+
+    vector_size_t numResultRows = 0;
+    while (auto result = task->next()) {
+      numResultRows += result->size();
+    }
+
+    folly::doNotOptimizeAway(numResultRows);
+  }
+
+  core::PlanFragment makeGroupByPlan(const std::string& aggregate) {
+    return PlanBuilder()
+        .tableScan(inputType_)
```

Comment:
Curious that if the aggregation reads data from table scan, wouldn't this benchmark measure the performance of aggregation in combination with table scan. Is this desired?

Reply:
Yes, this benchmark measures combined performance sum + table scan vs. reduce_agg + table scan. However, table scan is not a dominant part here. If it was, there would not be 60x difference. Hence, that's Ok.